# Importing Libraries

In [67]:
#Import Libraries
import pandas as pd
import numpy as np
import os

 # Reading CSV files

In [68]:
#set the path variable
path = r'C:\Users\Sreelakshmi\Desktop\Instacart Basket Analysis'

In [69]:
#Read orders.csv file into df_ords
df_ords = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'orders.csv'), index_col = False)

In [70]:
#Read products.csv file into df_prods
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'products.csv'), index_col = False)

In [71]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [72]:
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [73]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# Deleting columns

In [74]:
df_ords = df_ords.drop(columns = ['eval_set'])

In [75]:
df_ords.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# Frequency checking

In [76]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

# Changing a Variable’s Data Type

In [77]:
# setting the data type of order_id column to str
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [78]:
df_ords.describe()

user_id  order_number     order_dow  order_hour_of_day  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.029782e+05  1.715486e+01  2.776219e+00       1.345202e+01   
std    5.953372e+04  1.773316e+01  2.046829e+00       4.226088e+00   
min    1.000000e+00  1.000000e+00  0.000000e+00       0.000000e+00   
25%    5.139400e+04  5.000000e+00  1.000000e+00       1.000000e+01   
50%    1.026890e+05  1.100000e+01  3.000000e+00       1.300000e+01   
75%    1.543850e+05  2.300000e+01  5.000000e+00       1.600000e+01   
max    2.062090e+05  1.000000e+02  6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

# Renaming columns

In [79]:
#Renaming the column
df_ords.rename(columns = {'order_dow' : 'order_days_of_week'}, inplace = True)

In [80]:
df_ords.head()

order_id  user_id  order_number  order_days_of_week  order_hour_of_day  \
0  2539329        1             1                   2                  8   
1  2398795        1             2                   3                  7   
2   473747        1             3                   3                 12   
3  2254736        1             4                   4                  7   
4   431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# Importing Departments Dataset

In [81]:
df_dep = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'departments.csv'), index_col = False)

In [82]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

# Transposing

In [83]:
#Transposing df_dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [84]:
df_dep_t = df_dep.T

In [85]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# Adding Index

In [86]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [87]:
# creating new header by taking 1st row from the table
new_header = df_dep_t.iloc[0]

In [88]:
new_header

0    department
Name: department_id, dtype: object

In [89]:
# copy all the contents of df_dep_t from first row onwards
df_dep_t_new = df_dep_t[1:]

In [90]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [91]:
# Assigning the values of new_header to df_dep_t_new to the the headers
df_dep_t_new.columns = new_header

In [92]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# Creating Data Dictionary

In [93]:
# Transforming df_dep_t_new to a data dictionary
data_dict = df_dep_t_new.to_dict('index')

In [94]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [95]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [96]:
# To find out what the number 19 in department_id represents
print(data_dict.get('19'))

{'department': 'snacks'}


# Subsetting

In [97]:
# create a subset of df_prods which contains data only from Snacks department
df_snacks = df_prods[df_prods['department_id']==19]

In [98]:
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [99]:
#second method
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

In [100]:
df_snacks_2

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [101]:
#third method
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

In [102]:
df_snacks_3

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

# Step 2 from Task 4.4

In [103]:
df_ords.describe()

user_id  order_number  order_days_of_week  order_hour_of_day  \
count  3.421083e+06  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.029782e+05  1.715486e+01        2.776219e+00       1.345202e+01   
std    5.953372e+04  1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00  1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.139400e+04  5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.026890e+05  1.100000e+01        3.000000e+00       1.300000e+01   
75%    1.543850e+05  2.300000e+01        5.000000e+00       1.600000e+01   
max    2.062090e+05  1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

In [104]:
# setting the datatype of user_id  column to str
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [105]:
# setting the datatype of order_number  column to str
df_ords['order_number'] = df_ords['order_number'].astype('str')

In [106]:
df_ords.describe()

order_days_of_week  order_hour_of_day  days_since_prior_order
count        3.421083e+06       3.421083e+06            3.214874e+06
mean         2.776219e+00       1.345202e+01            1.111484e+01
std          2.046829e+00       4.226088e+00            9.206737e+00
min          0.000000e+00       0.000000e+00            0.000000e+00
25%          1.000000e+00       1.000000e+01            4.000000e+00
50%          3.000000e+00       1.300000e+01            7.000000e+00
75%          5.000000e+00       1.600000e+01            1.500000e+01
max          6.000000e+00       2.300000e+01            3.000000e+01

In [107]:
df_ords.head()

order_id user_id order_number  order_days_of_week  order_hour_of_day  \
0  2539329       1            1                   2                  8   
1  2398795       1            2                   3                  7   
2   473747       1            3                   3                 12   
3  2254736       1            4                   4                  7   
4   431534       1            5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

 # Step 3

In [108]:
# Already renamed the column order_dow to order_days_of_week.
# If we don't want to overwrite the dataframe then use the statement
# df_ords.rename(columns = {'order_dow' : 'order_days_of_week'}, inplace = False)

# Step 4

In [109]:
# finding busiest hour for placing orders
df_ords['order_hour_of_day'].value_counts(dropna = True)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

### Answer: The busiest hour for placing orders is at 10 am.

# Step 5

In [110]:
#Getting the value of 4 from data_dict
print(data_dict.get('4'))

{'department': 'produce'}


### Answer: The department_id 4 represents the department 'produce'.

# Step 6

In [111]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [112]:
# subsetting the dataframe which contains only breakfast sales info
df_breakfast = df_prods[df_prods['department_id'] == 14]

In [113]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

# Step 7

In [114]:
# subsetting items for dinner parties
df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

In [115]:
df_dinner_parties

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

# Step 8

In [116]:
# Last created data frame is df_dinner_parties
df_dinner_parties.shape

(7650, 5)

### Answer: There are 7650 rows in the dataframe df_dinner_parties

# Step 9

In [117]:
df_user1 = df_ords.loc[df_ords['user_id'] == '1']

In [118]:
df_user1

order_id user_id order_number  order_days_of_week  order_hour_of_day  \
0   2539329       1            1                   2                  8   
1   2398795       1            2                   3                  7   
2    473747       1            3                   3                 12   
3   2254736       1            4                   4                  7   
4    431534       1            5                   4                 15   
5   3367565       1            6                   2                  7   
6    550135       1            7                   1                  9   
7   3108588       1            8                   1                 14   
8   2295261       1            9                   1                 16   
9   2550362       1           10                   4                  8   
10  1187899       1           11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

# Step 10

In [119]:
df_user1.describe()

order_days_of_week  order_hour_of_day  days_since_prior_order
count           11.000000          11.000000               10.000000
mean             2.636364          10.090909               19.000000
std              1.286291           3.477198                9.030811
min              1.000000           7.000000                0.000000
25%              1.500000           7.500000               14.250000
50%              3.000000           8.000000               19.500000
75%              4.000000          13.000000               26.250000
max              4.000000          16.000000               30.000000

### Answer:
The User1 has placed a total of 11 orders.
Considering the day of the week 1 as Monday, the User1 has placed orders only from Mondays to Thursdays.
He hasn't placed any orders from Fridays to Sundays. More than 50% of his orders are placed on Wednesdays and Thursdays.
He places order on an average around 10 am.
Also he paces order on an average of 19 days since prior order.
It is also noted that he has placed two orders on the same day.

# Step 12

In [120]:
# Exporting df_ords dataframe as “orders_wrangled.csv”
df_ords.to_csv(os.path.join(path, 'Data','Prepared Data', 'orders_wrangled.csv'))

# Step 13

In [121]:
# Exporting df_dep_t_new  dataframe as “departments_wrangled.csv”
df_dep_t_new.to_csv(os.path.join(path, 'Data','Prepared Data', 'departments_wrangled.csv'))